pyoneer/rl/rollouts/gym_ops.py

```diff
@@ -0,0 +1,164 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+from gym import spaces
+
+import tensorflow as tf
+
+from pyoneer.debugging import debugging_ops
+from pyoneer.rl.wrappers import batch_impl
+
+
+def space_to_spec(space):
+    """Determines a tf.TensorSpec from the `gym.Space`.
+
+    Args:
+        space: a `gym.Space` instance (i.e. `env.action_space`)
+
+    Raises:
+        `TypeError` when space is not a `gym.Space` instance.
+
+    Returns:
+        Possibly nested `tf.TensorSpec`.
+    """
+    if space.__class__ in [spaces.Discrete, spaces.MultiDiscrete,
+                           spaces.MultiBinary, spaces.Box]:
+        return tf.TensorSpec(tf.TensorShape(space.shape),
+                             tf.dtypes.as_dtype(space.dtype))
+    elif isinstance(space, spaces.Tuple) or isinstance(space, spaces.Dict):
+        return tf.nest.map_structure(space_to_spec, space.spaces)
+    raise TypeError('`space` not supported: {}'.format(type(space)))
+
+
+class Transition(collections.namedtuple(
+        'Transition', ['state', 'reward', 'next_state', 'terminal', 'weight'])):
+
+    __slots__ = ()
```

Comment:
This seems like a micro-optimization and is sufficiently uncommon to require a comment at minimum. :)